core/src/main/java/org/opensearch/sql/expression/operator/arthmetic/ArithmeticFunction.java

```diff
@@ -43,168 +51,130 @@ public class ArithmeticFunction {
-            LONG, LONG, LONG),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprFloatValue(v1.floatValue() + v2.floatValue())),
-            FLOAT, FLOAT, FLOAT),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprDoubleValue(v1.doubleValue() + v2.doubleValue())),
-            DOUBLE, DOUBLE, DOUBLE)
-    );
+  private static SerializableFunction baseArithmeticFunction(SerializableBiFunction<BigDecimal, BigDecimal, BigDecimal> formula,
+                                                             Boolean secondArgumentZeroCheck,
+                                                             SerializableFunction<ExprValue, BigDecimal> argValue,
+                                                             SerializableFunction<BigDecimal, Number> convertValue,
+                                                             ExprCoreType dataType) {
+    return FunctionDSL.impl(
+            FunctionDSL.nullMissingHandling(
+                    (ExprValue v1, ExprValue v2) -> secondArgumentZeroCheck && v2.byteValue() == 0 ? ExprNullValue.of() :
+                            ExprValueUtils.fromObjectValue(convertValue.apply(formula.apply(argValue.apply(v1), argValue.apply(v2))))),
+            dataType, dataType, dataType);
   }
 
-  private static DefaultFunctionResolver subtract() {
-    return FunctionDSL.define(BuiltinFunctionName.SUBTRACT.getName(),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprByteValue(v1.byteValue() - v2.byteValue())),
-            BYTE, BYTE, BYTE),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprShortValue(v1.shortValue() - v2.shortValue())),
-            SHORT, SHORT, SHORT),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprIntegerValue(Math.subtractExact(v1.integerValue(),
-                    v2.integerValue()))),
-            INTEGER, INTEGER, INTEGER),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprLongValue(Math.subtractExact(v1.longValue(), v2.longValue()))),
-            LONG, LONG, LONG),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprFloatValue(v1.floatValue() - v2.floatValue())),
-            FLOAT, FLOAT, FLOAT),
-        FunctionDSL.impl(
-            FunctionDSL.nullMissingHandling(
-                (v1, v2) -> new ExprDoubleValue(v1.doubleValue() - v2.doubleValue())),
-            DOUBLE, DOUBLE, DOUBLE)
+  private static DefaultFunctionResolver baseArithmeticParser(SerializableBiFunction<BigDecimal, BigDecimal, BigDecimal> formula,
+                                                      Boolean secondArgumentZeroCheck, FunctionName functionName) {
```

Comment:
There's a difference between Boolean (Object) and boolean (simple data type).  If you're passing in true (simple data type) it needs to be boxed into a Boolean object (which is done implicitly) and we have a small performance impact.  

Use boolean here instead.

Reply:
As the performance hit from this base function was too big it was taken out